firestore/query.go

```diff
@@ -250,6 +251,159 @@ func (q *Query) processCursorArg(name string, docSnapshotOrFieldValues []interfa
 
 func (q Query) query() *Query { return &q }
 
+func getSafeCursorValue(vProto *pb.Value, q Query) (interface{}, error) {
+	switch v := vProto.ValueType.(type) {
+	case *pb.Value_ReferenceValue:
+		refVal := vProto.GetReferenceValue()
+		if strings.HasPrefix(refVal, q.path) {
+			refVal = refVal[len(q.path)+1:]
+		}
+		return refVal, nil
+	case *pb.Value_IntegerValue:
+		// TODO: this shouldn't be needed, but tests send things as int, this returns
+		// int64, so diff fails
+		return v.IntegerValue, nil //(v.IntegerValue), nil
+	default:
+		i, err := createFromProtoValue(vProto, q.c)
+		if err != nil {
+			q.err = err
+			return q, err
+		}
+		return i, nil
+	}
+}
+
+// FromProto creates a new Query object from a RunQueryRequest. This can be used
+// in combintation with ToProto to serialize Query objects.
+func (q Query) FromProto(pbQuery *pb.RunQueryRequest) (Query, error) {
+	// Ensure we are starting from an empty query, but with this client.
+	q = *q.c.Query()
+
+	// 	err                    error
+	q.err = nil
+
+	pbq := pbQuery.GetStructuredQuery()
+	if from := pbq.GetFrom(); len(from) > 0 {
+		// collectionID           string
+		q.collectionID = from[0].CollectionId
+		// allDescendants indicates whether this query is for all collections
+		// that match the ID under the specified parentPath.
+		q.allDescendants = from[0].AllDescendants
+	}
+
+	// 	path                   string // path to query (collection)
+	// 	parentPath             string // path of the collection's parent (document)
+	parent := pbQuery.GetParent()
+	q.parentPath = parent
+	q.path = parent + "/" + q.collectionID
+
+	// 	startVals, endVals     []interface{}
+	// 	startDoc, endDoc       *DocumentSnapshot
+	// 	startBefore, endBefore bool
+	if startAt := pbq.GetStartAt(); startAt != nil {
+		if startAt.GetBefore() {
+			q.startBefore = true
+		}
+		for _, v := range startAt.GetValues() {
+			c, err := getSafeCursorValue(v, q)
+			if err != nil {
+				q.err = err
+				return q, err
+			}
+
+			var newQ Query
+			if startAt.GetBefore() {
+				newQ = q.StartAt(c)
+			} else {
+				newQ = q.StartAfter(c)
+			}
+
+			q.startVals = append(q.startVals, newQ.startVals...)
+		}
+	}
+	if endAt := pbq.GetEndAt(); endAt != nil {
+		for _, v := range endAt.GetValues() {
+			c, err := getSafeCursorValue(v, q)
+			if err != nil {
+				q.err = err
+				return q, err
+			}
+
+			var newQ Query
+			if endAt.GetBefore() {
+				newQ = q.EndBefore(c)
+				q.endBefore = true
+			} else {
+				newQ = q.EndAt(c)
+			}
+			q.endVals = append(q.endVals, newQ.endVals...)
+
+		}
+	}
+
+	// 	selection              []FieldPath
+	if s := pbq.GetSelect(); s != nil {
+		fields := s.GetFields()
+		fieldStrings := make([]string, 0)
+
+		// As part of sending to protos, field paths are escaped. Unescape
+		for _, v := range fields {
+			fieldPath := v.GetFieldPath()
+			if strings.Contains(fieldPath, "`") {
+				fieldPath = strings.ReplaceAll(fieldPath, "`", "")
```

Comment:
Again, this would not be needed if we stored the Proto representation. This also seems to be missing the unescape on `\` (which also allows \\\` to be used to represent \`)

Reply:
Moved field path to proto rep.